### ISSUE TITLE
Install dotnet binary in a Linux pre-defined path

### ISSUE THREAD

**Issue by jpfreyen** [dotnet/core-setup#5221]
## Steps to reproduce
sudo dnf install dotnet-sdk-2.2

## Expected  behavior
After install, I can just type 'dotnet' and it will run!

## Actual behavior
terminal shell cannot find 'dotnet'
[jfreyensee@fedora-virtualfedora ~]$ dotnet
bash: dotnet: command not found..

## Environment data

dotnet installs to /usr/share/dotnet which is NOT found by the PATH variable (or the inherent default paths of a Linux terminal).  2 solutions here:

1.  rpm package creates a symbolic link /usr/bin/dotnet that points to /usr/share/dotnet/dotnet
2. rpm package installs to /usr/lib like java does then creates a /usr/bin/dotnet symbolic link pointing back to /usr/lib/dotnet/dotnet

I'm filing a new bug on this because I see this was a problem dating back to 2017, lots of ppl on Ubuntu complaining and the bug was closed with it obviously not being fixed because I'm having the problem again two years later, this time on a Fedora 28 box.

Come on guys, this is Linux 101, install binaries in default paths to make it easier on the consumer to use it!

**Comment by leecow** [comment]
Thanks for the report @jpfreyen and sorry for the difficulty. Fedora 28 is my daily driver distro at the moment and this really does work just fine in the vast majority of cases. The dotnet-host rpm is what is responsible for creating (and usually does) the symlink in /usr/bin - we'll need to see if we can figure out why this didn't happen on your system. 

Can you confirm the symlink was not created (ie `ls -l /usr/bin/dotnet`)? If it's not there, could you `sudo ln -s /usr/share/dotnet/dotnet /usr/bin/dotnet` then `dotnet --info` to see if that gets things working as expected?

**Comment by jpfreyen** [comment]
Yes, I can confirm the symlink wasn't created:

[jfreyensee@fedora-virtualfedora ~]$ ls -l /usr/bin/dotnet
ls: cannot access '/usr/bin/dotnet': No such file or directory

I wound up manually setting my PATH for now to try and get things working:

[jfreyensee@fedora-virtualfedora ~]$ which dotnet
/usr/share/dotnet/dotnet
[jfreyensee@fedora-virtualfedora ~]$ echo $PATH
/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/usr/share/dotnet:/home/jfreyensee/.dotnet/tools:/opt/paket:/home/jfreyensee/.local/bin:/home/jfreyensee/bin:/opt/paket

[jfreyensee@fedora-virtualfedora ~]$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.2.104
 Commit:    73f036d4ac

Runtime Environment:
 OS Name:     fedora
 OS Version:  28
 OS Platform: Linux
 RID:         fedora.28-x64
 Base Path:   /usr/share/dotnet/sdk/2.2.104/

Host (useful for support):
  Version: 2.2.2
  Commit:  a4fd7b2c84

.NET Core SDKs installed:
  2.2.104 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
[jfreyensee@fedora-virtualfedora ~]$

**Comment by dagood** [comment]
I looked into this some, and unfortunately I'm not able to repro or find any general reasons our postinstall scriptlet might be failing in your case. For some background, you can see the way our package tries to set up the symlink like this:

```
#> rpm -q --scripts dotnet-host

postinstall scriptlet (using /bin/sh):
#!/bin/sh
#
# Copyright (c) .NET Foundation and contributors. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.
#
echo "Creating dotnet host symbolic link: /usr/bin/dotnet"
ln -sf "/usr/share/dotnet/dotnet" "/usr/bin/dotnet"
postuninstall scriptlet (using /bin/sh):
#!/bin/sh
#
# Copyright (c) .NET Foundation and contributors. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.
#
# Run the script only when newer version is not installed. Skip during package upgrade
if [ $1 = 0 ]; then
   echo "Removing dotnet host symbolic link"
   unlink /usr/bin/dotnet
fi
```

Here's a simple Dockerfile that exercises this install flow, based on https://dotnet.microsoft.com/download/linux-package-manager/fedora28/sdk-current:

```Dockerfile
FROM fedora:28

RUN dnf install -y sudo wget

RUN sudo rpm --import https://packages.microsoft.com/keys/microsoft.asc \
  && wget -q https://packages.microsoft.com/config/fedora/27/prod.repo \
  && sudo mv prod.repo /etc/yum.repos.d/microsoft-prod.repo \
  && sudo chown root:root /etc/yum.repos.d/microsoft-prod.repo

RUN sudo dnf install -y dotnet-sdk-2.2

RUN ls -l /usr/bin/dotnet \
  && whereis dotnet \
  && dotnet --info

# Put this in a file called `Dockerfile` in an empty directory, then run `docker build .`.
```

I uploaded the `/var/log/dnf.log` that ends up in the container: https://gist.github.com/dagood/3f297d700e497c2c65218236ae601fbd.

@jpfreyen, maybe it would be helpful to compare against my results above and see if something is different on your machine? Maybe my "from scratch" Dockerfile isn't the same as your install story--have you installed earlier versions of `dotnet-sdk-*` before?

I'm new to Linux package managers, so I'm no doubt missing some ways to troubleshoot. @omajid, do you think you can you help out with this?

**Comment by omajid** [comment]
I don't have any immediate ideas about troubleshooting, but I am wondering if we can remove the rpm scriptlets and make (a part of the) installation more robust. Would it be possible to add a symlink file to the RPM file itself?

In the Fedora .NET Core SIG version of .NET Core 2.2, we make `/usr/bin/dotnet` a symlink owned/installed by the RPM: https://pagure.io/dotnet-sig/dotnet-2-2/blob/2b0e085e5507f03e13bd283e60b4fc3bf6e30773/f/dotnet2.2.spec#_312 (just like `/usr/share/dotnet/dotnet` is). That avoid any need for rpm scriptlets which I have always found a little tricky.

**Comment by jpfreyen** [comment]
@dagood I did have previous versions of dotnet installed on my machine.  

Maybe postinstall isn't the best place, and say, %install is better.  I did a quick look at Java's spec file, and all their symbolic links are in the %install section (though it looks like they are just resetting their .jar's there).

https://github.com/puzzle/java-oracle-rpm/blob/master/java-1.7.0-oracle.spec

I think that is slightly eluding to what @omajid said, try setting the symlink somewhere outside a scriptlet.

**Comment by dagood** [comment]
Case in point for scriptlets not being easy to put together, the original Core-Setup RPM package builds had a bug where the symlink was being deleted by the postuninstall scriptlet because the `if [ $1 = 0 ];` was missing. There are no git tags containing the initial commit (https://github.com/dotnet/core-setup/commit/7ac2673afb55937a702ae770ad020a28a43221cf) that don't contain the fix commit (https://github.com/dotnet/core-setup/commit/e034da2ec2bc03a319be36237276570a5842bda2), so it's pretty unlikely that's the cause of the problem in this thread. (It looks like you would have had to upgrade from a 2.0.0 preview daily build, which was a while ago.)

We don't use spec files in core-setup, but I think I see how to do the same thing with what we have. I'll give it a shot. I don't understand the Core-Setup Debian package build very well yet, but it looks like we do it the better way there already.

**Comment by jpfreyen** [comment]
I did upgrade from a 2.1.0 preview build.

As an experiment, maybe you can use alien and convert the .deb to an .rpm and see how well the .deb solution holds up on an .rpm?  I haven't played with alien in a long, long time.

**Comment by dagood** [comment]
Interesting, hadn't seen that before. Something we're already considering though is consolidating on FPM, which is what we use for RPM packaging. Our Debian packages are created using a custom tool that would be nice to stop using in favor of a maintained project. 😄

**Comment by dagood** [comment]
Submitted https://github.com/dotnet/core-setup/pull/5380 to remove the scriptlets. That said... we haven't made much progress on diagnosing the original issue. @jpfreyen, have you had a chance to poke at your dnf logs to see if anything stands out?

**Comment by dagood** [comment]
Closing for now: this particular instance of the problem needs more info to dig in any further. Please let me know if you find anything else.

This issue did result in improving the RPM packages, though, so hopefully strange issues like this will be rarer in the future. Thanks! 😄

**Comment by jpfreyen** [comment]
Sorry, I forgot to respond. I found the original dotnet install log I can attach it below for review (there is a whole lot there and I did a lot, installing dotnet.2.2 and then getting rid of older packages):

[dnf-dotnet.log](https://github.com/dotnet/core-setup/files/2984770/dnf-dotnet.log)

But I'm glad this exercise did in fact improve the RPM package :-)

**Comment by dagood** [comment]
No problem, thanks for digging up the log. I took a look, and it looks like your upgrade was from 2.1.5 to 2.2.2:

```
2019-02-20T19:54:50Z DEBUG ---> Package dotnet-host.x86_64 2.1.5-1.fc28 will be upgraded
2019-02-20T19:54:50Z DEBUG ---> Package dotnet-host.x86_64 2.2.2-1 will be an upgrade
```

The `.fc28` catches my eye... that's a Fedora SIG package, right @omajid?

If so, that would explain it! The SIG package installs the symlink, then when upgrading to the Microsoft package, the symlink is removed because the Msft package doesn't "own" it. So it turns out we've fixed the root cause by making the Msft package also install a symlink. Hurray! 😄 

(The fix will be in the next released rpms for 2.1, 2.2, and 3.0.)
